Account- and product-level issues

layoutd opened the pull request:
### Changes proposed in this Pull Request:

Closes #154.

This adds the REST endpoint for retrieving the account- and product-level issues for a Merchant Center account.
- One-ish endpoint is added: `GET /gla/mc/issues` retrieves all issues. 
    - However, adding `/account` or `/product` will filter by type (e.g., `/gla/mc/issues/product`)
    - If no type filter is specified, account-level issues will be listed first 
    - Product-level issues are grouped by description (so `Missing value [color]` could have one or more `applicable_countries`) and sorted by WooCommerce product ID.
    - Only issues for products submitted by GLA will be returned ([filtered using the expected Google offer ID format](https://github.com/woocommerce/google-listings-and-ads/pull/421/files#diff-0db42cbfdccd143bd421943def87d36d0e12e3f1514bf0e05c0cc27c6a658581R149))
- Several query params are also available
    - The `page` and `per_page` parameters can be used for pagination (https://developer.wordpress.org/rest-api/using-the-rest-api/pagination/)

The issues are retrieved from the Google API and cached in a transient with a one-hour lifetime. The full list of issues is generated by:
 - One call to [`accountstatuses.get`](https://developers.google.com/shopping-content/reference/rest/v2.1/accountstatuses/get), and then 
- As many calls to [`productstatuses.list`](https://developers.google.com/shopping-content/reference/rest/v2.1/productstatuses/get) with a `maxResults` of `200` as necessary

The information that's cached and returned with the response is considerably pared down:
- Issue `type` - `account` or `product`
- Associated `product` (title or `All products` for account-level issues)
- Description of the issue (`issue`)
- Google's internal `code` in case further filtering is necessary. 
    - For example, it appears that `policy_enforcement_account_disapproval` means account suspension, but is returned for every product. Not much can be done by sending the user to edit their WooCommerce product.
- Recommended action and link to the relevant documentation (`action` and `action_link`)
- For products, groups by `issue` and shows all `applicable_countries` and includes the `product_id`
- Response includes values for `total` number of issues, and `page` to assist with pagination

<details>
<summary><em>See example REST response</em></summary>

```json
{
  "issues": [
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "policy_enforcement_account_disapproval",
      "issue": "Suspended account for policy violation",
      "action": "Remove products that violate our policies, or request a manual review",
      "action_link": "https://support.google.com/merchants/answer/2948694",
      "applicable_countries": [
        "ES"
      ]
    },
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "invalid_currency_for_country",
      "issue": "Unsupported currency",
      "action": "Use the currency of the country of sale in your product data",
      "action_link": "https://support.google.com/merchants/answer/160637",
      "applicable_countries": [
        "KR"
      ]
    },
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "missing_shipping",
      "issue": "Missing value [shipping]",
      "action": "Add shipping costs for your product",
      "action_link": "https://support.google.com/merchants/answer/6239383",
      "applicable_countries": [
        "AR", "AT", "AU", "BE", "BH", "BR", "BY", "CA", "CH", "CL", "CO", "CR", "CZ", "DE", "DK", "EC", "EG", "ES", "FI", "FR", "GE", "GR", "HK", "HU", "ID", "IE", "IL", "IN", "IT", "JO", "JP", "KR", "KW", "KZ", "LB", "MX", "MY", "NL", "NO", "NZ", "OM", "PE", "PH", "PL", "PT", "PY", "RO", "RU", "SE", "SG", "SK", "TR", "TW", "ZA"
      ]
    },
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "missing_item_attribute_for_product_type",
      "issue": "Missing value [size]",
      "action": "Add this attribute to your product data",
      "action_link": "https://support.google.com/merchants/answer/9762223",
      "applicable_countries": [
        "BR", "DE", "FR", "GB", "JP", "US"
      ]
    },
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "url_does_not_match_homepage",
      "issue": "Mismatched domains [link]",
      "action": "Use the same domain for product landing page URLs as in your Merchant Center website setting",
      "action_link": "https://support.google.com/merchants/answer/160050",
      "applicable_countries": [
        "AE", "AR", "AT", "AU", "BE", "BH", "BR", "BY", "CA", "CH", "CL", "CO", "CR", "CZ", "DE", "DK", "EC", "EG", "ES", "FI", "FR", "GB", "GE", "GR", "HK", "HU", "ID", "IE", "IL", "IN", "IT", "JO", "JP", "KR", "KW", "KZ", "LB", "MX", "MY", "NL", "NO", "NZ", "OM", "PE", "PH", "PL", "PT", "PY", "RO", "RU", "SE", "SG", "SK", "TR", "TW", "US", "UY", "UZ", "ZA"
      ]
    },
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "missing_business_registration_number",
      "issue": "Missing business registration number",
      "action": "Add your business registration number to this Merchant Center account",
      "action_link": "https://support.google.com/merchants/answer/9037766",
      "applicable_countries": [
        "KR"
      ]
    },
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "missing_item_attribute_for_product_type",
      "issue": "Missing value [age group]",
      "action": "Add this attribute to your product data",
      "action_link": "https://support.google.com/merchants/answer/9762223",
      "applicable_countries": [
        "BR", "DE", "FR", "GB", "JP", "US"
      ]
    },
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "missing_item_attribute_for_product_type",
      "issue": "Missing value [gender]",
      "action": "Add this attribute to your product data",
      "action_link": "https://support.google.com/merchants/answer/9762223",
      "applicable_countries": [
        "BR", "DE", "FR", "GB", "JP", "US"
      ]
    },
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "image_link_internal_error",
      "issue": "Processing failed [image link]",
      "action": "Wait for the product image to be crawled again (up to 3 days)",
      "action_link": "https://support.google.com/merchants/answer/6240184",
      "applicable_countries": [
        "AE", "AR", "AT", "AU", "BE", "BH", "BR", "BY", "CA", "CH", "CL", "CO", "CR", "CZ", "DE", "DK", "EC", "EG", "ES", "FI", "FR", "GB", "GE", "GR", "HK", "HU", "ID", "IE", "IL", "IN", "IT", "JO", "JP", "KR", "KW", "KZ", "LB", "MX", "MY", "NL", "NO", "NZ", "OM", "PE", "PH", "PL", "PT", "PY", "RO", "RU", "SE", "SG", "SK", "TR", "TW", "US", "UY", "UZ", "ZA"
      ]
    },
    {
      "type": "product",
      "product": "Long Sleeve Tee",
      "product_id": 22,
      "code": "missing_item_attribute_for_product_type",
      "issue": "Missing value [color]",
      "action": "Add this attribute to your product data",
      "action_link": "https://support.google.com/merchants/answer/9762223",
      "applicable_countries": [
        "BR", "DE", "FR", "GB", "JP", "US"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "policy_enforcement_account_disapproval",
      "issue": "Suspended account for policy violation",
      "action": "Remove products that violate our policies, or request a manual review",
      "action_link": "https://support.google.com/merchants/answer/2948694",
      "applicable_countries": [
        "ES"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "invalid_currency_for_country",
      "issue": "Unsupported currency",
      "action": "Use the currency of the country of sale in your product data",
      "action_link": "https://support.google.com/merchants/answer/160637",
      "applicable_countries": [
        "KR"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "missing_shipping",
      "issue": "Missing value [shipping]",
      "action": "Add shipping costs for your product",
      "action_link": "https://support.google.com/merchants/answer/6239383",
      "applicable_countries": [
        "AR", "AT", "AU", "BE", "BH", "BR", "BY", "CA", "CH", "CL", "CO", "CR", "CZ", "DE", "DK", "EC", "EG", "ES", "FI", "FR", "GE", "GR", "HK", "HU", "ID", "IE", "IL", "IN", "IT", "JO", "JP", "KR", "KW", "KZ", "LB", "MX", "MY", "NL", "NO", "NZ", "OM", "PE", "PH", "PL", "PT", "PY", "RO", "RU", "SE", "SG", "SK", "TR", "TW", "ZA"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "missing_item_attribute_for_product_type",
      "issue": "Missing value [size]",
      "action": "Add this attribute to your product data",
      "action_link": "https://support.google.com/merchants/answer/9762223",
      "applicable_countries": [
        "BR", "DE", "FR", "GB", "JP", "US"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "url_does_not_match_homepage",
      "issue": "Mismatched domains [link]",
      "action": "Use the same domain for product landing page URLs as in your Merchant Center website setting",
      "action_link": "https://support.google.com/merchants/answer/160050",
      "applicable_countries": [
        "AE", "AR", "AT", "AU", "BE", "BH", "BR", "BY", "CA", "CH", "CL", "CO", "CR", "CZ", "DE", "DK", "EC", "EG", "ES", "FI", "FR", "GB", "GE", "GR", "HK", "HU", "ID", "IE", "IL", "IN", "IT", "JO", "JP", "KR", "KW", "KZ", "LB", "MX", "MY", "NL", "NO", "NZ", "OM", "PE", "PH", "PL", "PT", "PY", "RO", "RU", "SE", "SG", "SK", "TR", "TW", "US", "UY", "UZ", "ZA"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "missing_business_registration_number",
      "issue": "Missing business registration number",
      "action": "Add your business registration number to this Merchant Center account",
      "action_link": "https://support.google.com/merchants/answer/9037766",
      "applicable_countries": [
        "KR"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "missing_item_attribute_for_product_type",
      "issue": "Missing value [age group]",
      "action": "Add this attribute to your product data",
      "action_link": "https://support.google.com/merchants/answer/9762223",
      "applicable_countries": [
        "BR", "DE", "FR", "GB", "JP", "US"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "missing_item_attribute_for_product_type",
      "issue": "Missing value [gender]",
      "action": "Add this attribute to your product data",
      "action_link": "https://support.google.com/merchants/answer/9762223",
      "applicable_countries": [
        "BR", "DE", "FR", "GB", "JP", "US"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "image_link_internal_error",
      "issue": "Processing failed [image link]",
      "action": "Wait for the product image to be crawled again (up to 3 days)",
      "action_link": "https://support.google.com/merchants/answer/6240184",
      "applicable_countries": [
        "AE", "AR", "AT", "AU", "BE", "BH", "BR", "BY", "CA", "CH", "CL", "CO", "CR", "CZ", "DE", "DK", "EC", "EG", "ES", "FI", "FR", "GB", "GE", "GR", "HK", "HU", "ID", "IE", "IL", "IN", "IT", "JO", "JP", "KR", "KW", "KZ", "LB", "MX", "MY", "NL", "NO", "NZ", "OM", "PE", "PH", "PL", "PT", "PY", "RO", "RU", "SE", "SG", "SK", "TR", "TW", "US", "UY", "UZ", "ZA"
      ]
    },
    {
      "type": "product",
      "product": "Polo",
      "product_id": 23,
      "code": "missing_item_attribute_for_product_type",
      "issue": "Missing value [color]",
      "action": "Add this attribute to your product data",
      "action_link": "https://support.google.com/merchants/answer/9762223",
      "applicable_countries": [
        "BR", "DE", "FR", "GB", "JP", "US"
      ]
    }
  ],
  "total": 20,
  "page": 1
}
```
</details>

**To Do**
- Standardized responses now ~Decide whether to reduce the amount of information cached and returned for each issue~
- Added in 00acecd33fcfaaa6968eaf48a52f8ccefed952f4 ~Include a link to the WooCommerce product using the MC product ID.~
- Only use/store/return the product ID and let the front end [generate the link](https://github.com/woocommerce/google-listings-and-ads/blob/8a88ef9b5675368828a4fe13c39f86713ce427e9/js/src/components/edit-product-link/index.js#L7-L14). ~Return product ID instead of `edit_link` to reduce cache size. Does the front end have a method like [`get_edit_post_link()`](https://developer.wordpress.org/reference/functions/get_edit_post_link/)?~
- Test with a real catalogue to find edge cases
    - Done a little bit, certainly other things will pop up.
- Use a custom DB table instead of a transient - **Separate PR**


### Detailed test instructions:

1. Connect a Merchant Center account (using the onboarding wizard) and sync some products.
2. Wait a little while for the products to appear in the Merchant Center (https://merchants.google.com/mc/items)
3. Make a request to `GET /gla/mc/issues`
    - Confirm that the response has the format detailed above.
    - Confirm the transient was created and the issues cached correctly:
    ```sql
    SELECT * FROM `wp_options` WHERE `option_name` = '_transient_gla_mc_issues'
    ```
    - Confirm the pagination values
4. Test other options:
    - Make a request to `GET /gla/mc/issues/account` and `GET /gla/mc/issues/product` to see results filtered by type
    - Include the `per_page` and/or `page` params to reduce and navigate through the results.
- Bonus: check results against the Shopping Content API: 
    https://developers.google.com/shopping-content/reference/rest/v2.1/accountstatuses/get , and 
    https://developers.google.com/shopping-content/reference/rest/v2.1/productstatuses/list
- Double bonus: manually create a product and ensure that it isn't included in the issues response (since it won't have the correct product ID prefix).

layoutd committed:
Endpoint for Merchant Center issues

layoutd committed:
Unused imports

layoutd committed:
Add product-level issues to the issues list.

layoutd committed:
Small fixes for product status retrieval
array merge, 200 maxresults and array values after filter

layoutd committed:
1 hours lifetime

layoutd committed:
Standardize issue data in controller

layoutd committed:
Include link to the affected product when possible

layoutd committed:
Add method to extract WC product ID from Google product ID

layoutd committed:
Only cache and return GLA-synced products

layoutd committed:
Only extract correctly prefixed product IDs

layoutd committed:
Add get_edit_post_link to WP proxy

layoutd committed:
Include underscore in Google Product ID extraction regex

layoutd committed:
Use WP proxy for get_edit_post_link

layoutd committed:
Protected instead of private

layoutd committed:
Unused import

layoutd committed:
Include return type

mikkamp reviewed (commented):
It's working good and I'm getting the results as expected. Although I'm a bit worried with the transient getting out of control for sites with large product sets / many errors. If the rest response is considerably pared down compared to the transient. Then why don't we convert earlier and save the smaller version? It can still be filtered by type in the same way. Or do we need to consider only retrieving a limit and implement the usage of next page token? Or do we prefer to get all results and handle the pagination for display only?

In the figma it seems to indicate that we are only showing a subset of products:
![image](https://user-images.githubusercontent.com/11388669/113892326-caaed280-97bd-11eb-9b16-b0424aa3fa2d.png)

Are we filtering the issues based on the same subset? Do we need to include filter by product id?

layoutd committed:
Change controller name for consistency

layoutd committed:
Typo

layoutd committed:
Fix edit post link context

layoutd committed:
Avoid returning an associative array

layoutd committed:
Preg_quote for slug

layoutd committed:
Reduce information stored in issues transient

- Only one of each type of error per product (regardless of number of geos)
- Convert issue data to reduced format before storage

layoutd committed:
Add pagination for issues

layoutd committed:
Include applicable_countries for product-level issues

layoutd committed:
Include product-issue documentation URL and issue codes

layoutd committed:
Endpoint to search by product issues by name or ID

layoutd committed:
Disallow search with GET or accounts

layoutd committed:
User query params for issue search

layoutd committed:
Sort product issues by product ID

layoutd committed:
Alphabetize applicable countries before caching

layoutd committed:
Extract product issues retrieval step

mikkamp reviewed (commented):
It works well, and I wasn't able to find any blocking errors. Just added thoughts on how we could standardize the API endpoints a bit more.

layoutd committed:
Unnecessary cast

layoutd committed:
Remove per_page from REST response

layoutd committed:
Prefix MC classes with MerchantCenter (if prefixed)

layoutd committed:
Better schema for the issues in the response

layoutd committed:
Extract get_account_issues and simplify refresh

layoutd committed:
Validation/sanitization for pagination and filter/search args

layoutd committed:
Add WC product ID to the cache and response

layoutd committed:
Remove product query param support

* Not actually included in the design

layoutd committed:
Don't include product edit link in response or cache
